### PR TITLE
Crank runner for Helix

### PR DESCRIPTION
**Pull request by caaavik-msft**
This PR contains initial work around making it so that devs inside Microsoft can use crank to schedule performance runs on Helix machines locally, rather than having to do them through CI pipelines only.

Currently, a lot of the logic around running performance tests are stored in a mix of Azure Pipelines yaml files, shell scripts, powershell scripts, python scripts, MSBuild project files across both the dotnet/runtime and dotnet/performance repositories. This PR attempts to move as much of that logic as possible into python scripts so that they can be run outside the context of a CI pipeline, and eventually we can update our CI pipelines to run these python scripts instead.

I also took this opportunity to fix up our python scripts to have proper type hints set, as well as configuring some settings so that we get good autocompletion and linting inside VSCode.

I have set up a crank agent that runs inside corpnet, which those inside Microsoft on a corpnet-connected machine will be able to use. For Microsoft employees working remotely, they can either use a DevBox (which has access to corpnet), or they can talk to a member of the .NET Performance Team to get an Azure Relay connection string which will allow communication outside corpnet.

An example of using crank to schedule a performance run on a remote machine is:
```
crank --config .\helix.yml --scenario micro --profile win-x64 --variable bdnArgs="--filter *Linq*" --profile msft-internal --variable buildNumber="caaavik-20230804.2"
```
This will run all the Linq microbenchmarks on a win-x64 helix machine against a local pre-built corerun. The results will be returned back and displayed in the terminal, it may also be uploaded to the Azure Data Explorer to view there.

### CODE REVIEW

**Comment by caaavik-msft** [comment]
Some things that still need fixing before this can be merged:
- I need to add some documentation on how to use this
- There are a few bugs left with running scenarios I need to fix
- I also don't have support just yet for getting the scenarios to send the results back to crank, they will be uploaded to ADX though.
- I had to add an exception for `.git` in the `.gitignore` otherwise it would not get uploaded to the crank agent where we run git commands

File: scripts/ci_setup.py
Comment (caaavik-msft) [inline scripts/ci_setup.py:361]: Reminder to uncomment this and figure out what to do for commit time for local runs.

File: eng/performance/benchmark_jobs.yml
Comment (LoopedBard3) [inline eng/performance/benchmark_jobs.yml:126]: It looks like we do still use this: https://github.com/dotnet/performance/blob/main/eng/performance/benchmark_jobs.yml#L99

File: eng/performance/helix_sender.Dockerfile
Comment (cincuranet) [inline eng/performance/helix_sender.Dockerfile:1]: I would vote for `3.10-bookworm` (or maybe better `3.10-slim-bookworm`), to be on current Debian 12.

File: scripts/run_performance_job.py
Comment (cincuranet) [inline scripts/run_performance_job.py:182]: This needs to be synced with https://github.com/dotnet/runtime/pull/90093 and https://github.com/dotnet/runtime/pull/90113.

File: eng/performance/benchmark_jobs.yml
Comment (LoopedBard3) [inline eng/performance/benchmark_jobs.yml:126]: I am not finding _PerfLabArguments anywhere else so I think the rename got all the updates. We can probably remove this.

File: eng/performance/benchmark_jobs.yml
Comment (LoopedBard3) [inline eng/performance/benchmark_jobs.yml:133]: Since we install python in at least 2 places (here and scnearios.yml), it may be beneficial to see if we can pull the version into a single variable or version in a file so we only have to update it in one place.

File: scripts/performance_setup.py
Comment (LoopedBard3) [inline scripts/performance_setup.py:239]: I think we should either be just the runtime commit time, or both the runtime and performance times. I prefer to have both, even if the performance repo commit time gets added as additional data.

File: eng/performance/benchmark_jobs.yml
Comment (caaavik-msft) [reply eng/performance/benchmark_jobs.yml]: I've changed it now so that we just use python 3.6 actually, but dataclasses are supported by installing the dataclasses pip package

File: scripts/performance_setup.py
Comment (caaavik-msft) [reply scripts/performance_setup.py:239]: At the moment it's only practical to get the performance repo commit time, as only the core root folder is uploaded to the crank agent. To get the runtime commit information we may need to look at sending up the `.git` directory somehow to crank. I'll leave this TODO in here since I don't think I will be able to resolve it just yet.

File: eng/performance/helix_sender.Dockerfile
Comment (caaavik-msft) [reply eng/performance/helix_sender.Dockerfile]: I tested out slim-bookworm but it was causing too many issues with other dependencies such as wget so just using bookworm now.

File: helix.yml
Comment (caaavik-msft) [reply helix.yml]: I have removed this setting now, it's only used if crank builds a .NET program which we aren't doing inside crank.

File: scripts/run_performance_job.py
Comment (caaavik-msft) [reply scripts/run_performance_job.py:182]: Done

File: eng/performance/benchmark_jobs.yml
Comment (caaavik-msft) [reply eng/performance/benchmark_jobs.yml]: Removed

File: scripts/performance_setup.py
Comment (LoopedBard3) [reply scripts/performance_setup.py:239]: Sounds good to me.

File: eng/performance/benchmark_jobs.yml
Comment (LoopedBard3) [reply eng/performance/benchmark_jobs.yml]: Is dataclasses something we should add to the requirements.txt file?

File: eng/performance/helix.proj
Comment (DrewScoggins) [inline eng/performance/helix.proj:5]: It seems like we are setting the HelixPreCommands here and then resetting them a few lines down. Am I missing something here?

File: scripts/dotnet.py
Comment (DrewScoggins) [inline scripts/dotnet.py:925]: Do we just no longer need this argument?

File: eng/performance/helix.proj
Comment (caaavik-msft) [reply eng/performance/helix.proj:5]: This is with UseCoreRun == true set which is true for runs against a local runtime build, the reason for this is that in the runtime repo we set the precommands in the helix.proj, but in the performance repo we set them in the yaml file. So this was a way to merge the two .proj files. In the future I think consolidating it so that more of this is done inside the python scripts would be good.

File: scripts/dotnet.py
Comment (caaavik-msft) [reply scripts/dotnet.py:925]: Yep, I checked and this argument was not used anywhere

File: scripts/performance_setup.py
Comment (DrewScoggins) [reply scripts/performance_setup.py:239]: Sorry to unresolve, but I am confused here. Shouldn't we be getting the commit time from calling ci_setup.py? Specifically, we call `dotnet.get_commit_time` and this properly gets the commit based on what repo we are running against.

File: scripts/run_performance_job.py
Comment (DrewScoggins) [inline scripts/run_performance_job.py:1]: Not sure exactly where to put this comment, but do we actually support running against WASM as this currently stands? I am having a hard time understanding where we package up the various runtime packs and such. Is that something that just needs to be handled by the user?

File: scripts/performance_setup.py
Comment (DrewScoggins) [reply scripts/performance_setup.py:239]: Oh, I see I missed that we don't follow that code path here because we are running against something local. It does feel like this is something we could grab locally when we send the job off to crank, but that feels outside the scope for sure.

File: scripts/run_performance_job.py
Comment (caaavik-msft) [reply scripts/run_performance_job.py:1]: no it doesn't support wasm right now, I have a TODO for it in the code but I need to understand more about what all the inputs are and where it downloads everything

Comment (caaavik-msft) [reply scripts/run_performance_job.py:1]: I know there are some files in the runtime repo that are used in wasm runs, and so to support that we would need to add additional sources to the helix.yml which ensures those directories are sent to crank

**Review by DrewScoggins — APPROVED** [review]
LGTM. Obviously there is a lot here, and I would be lying if I said that I now fully understand all of the intricacies. Overall, the change looks good though and I am confident that this shouldn't cause any breakage in our existing data pipelines.

File: docs/crank-to-helix-workflow.md
Comment (DrewScoggins) [inline docs/crank-to-helix-workflow.md:32]: Maybe also include additional configurations here (Mono, WASM, iOS, Android, etc...)

File: docs/crank-to-helix-workflow.md
Comment (DrewScoggins) [inline docs/crank-to-helix-workflow.md:60]: This command does not work in Powershell in Windows Terminal. I had to replace the `&&` with `;`

File: docs/crank-to-helix-workflow.md
Comment (DrewScoggins) [inline docs/crank-to-helix-workflow.md:47]: Do we currently have an alternative?

Comment (caaavik-msft) [reply docs/crank-to-helix-workflow.md:47]: Yes, using Azure Relay, didn't want to go into it in too much detail in the docs.

File: docs/crank-to-helix-workflow.md
Comment (LoopedBard3) [inline docs/crank-to-helix-workflow.md:47]:
```suggestion
  - If you are unable to use DevBox and can't get corpnet access, then please email [dotnetperf@microsoft.com](mailto:dotnetperf@microsoft.com) so that we can give you an alternative.
```

File: docs/crank-to-helix-workflow.md
Comment (DrewScoggins) [reply docs/crank-to-helix-workflow.md:47]: OK, I wasn't sure.

File: docs/crank-to-helix-workflow.md
Comment (DrewScoggins) [inline docs/crank-to-helix-workflow.md:45]: Maybe we can make this a little clearer. It is not immediately obvious that your computer must both be connected to the ethernet in a MS building and connected to the VPN. Maybe we should also start with saying that people should use Devbox, just because I wonder how many people actually have a computer they use day-to-day that meets the requirements.

File: docs/crank-to-helix-workflow.md
Comment (DrewScoggins) [inline docs/crank-to-helix-workflow.md:97]: Maybe it is also worth adding here that using `--profiles` multiple times is additive and does not destroy what was inserted with a previous call. That is a fairly unintuitive way for a command line option to work.

File: docs/crank-to-helix-workflow.md
Comment (DrewScoggins) [inline docs/crank-to-helix-workflow.md:118]: Might it be worth it to add a query here that allows people to easily compare their results against some build we already have in the data?

Comment (caaavik-msft) [reply docs/crank-to-helix-workflow.md:131]: I was going to include the one you sent me, but I think for now I'm not sure if this is the right way to do comparisons. Not going to block the PR on this.

**Review by cincuranet — APPROVED** [review]

**Pull request merged**
